[examples] Update catvdog for Klio v2

econchick opened the pull request:
<!--- Describe your changes 

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
-->

**Note**: I tried to update httplib2 (re: [this alert](https://github.com/spotify/klio/network/alert/examples/catvdog/job-requirements.txt/httplib2/open)) but it’s actually hard-pinned in [apache-beam](https://github.com/apache/beam/blob/master/sdks/python/setup.py#L150).

Also, we have yet to do a write up for this example for the docs site, but it is on our backlog.

I confirmed this works by running `klio job test` and `klio job run --direct-runner` + publishing a message to it.

<!--- How have you tested this?
Some valid responses are:

* "I have included unit tests" 
* "I have included integration tests"
* "I successfully ran my jobs with this code"

-->

## Checklist for PR author(s)
<!-- If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing left to do. -->
- [ ] Format the pull request title like `[cli] Fixes bugs in 'klio job fake-cmd'`.
- [ ] Changes are covered by unit tests (no major decrease in code coverage %) and/or integration tests.
- [ ] Document any relevant additions/changes in the appropriate spot in `docs/src`.


<!---
for more information on how to submit valuable contributions,
see https://opensource.guide/how-to-contribute/#how-to-submit-a-contribution
-->

econchick commented:
Ah I forgot: I was able to address [this dependabot issue](https://github.com/spotify/klio/network/alert/examples/catvdog/job-requirements.txt/tensorflow/open).

DanSimon commented on examples/catvdog/README.rst:
```diff
@@ -0,0 +1,52 @@
+
+.. note::
+
+    Make sure the minor version matches with the base Docker image stated on the ``FROM`` line in the `Dockerfile <./Dockerfile#L2>`_.
```
Is this referring to the minor version on the user's local dev environment, or something else?

econchick replied:
Yeah - is there a way that I could make that a bit more clear?

DanSimon replied:
Actually that raises another question...is that requirement actually necessary?  Isn't the reason we have distinct `klio-cli` and `klio-exec` so that the minor python versions can differ?

econchick replied:
actually yeah you're right lol will update

DanSimon commented on examples/catvdog/README.rst:
```diff
@@ -0,0 +1,52 @@
+
+    # Create a klio-job.yaml based off of the example config
+    $ export GCP_PROJECT=<your gcp project>
+    $ sed -e "s/\$GCP_PROJECT/${GCP_PROJECT}/" \
```
just a thought: I wonder if we could incorporate this functionality into the `klio job config set` command, like adding functionality to permanently apply a template parameter.  Could be useful but maybe not worth the effort.

econchick replied:
ah hmm yeah I like that idea from a user perspective, but that could tightly couple some things (`pipeline_options.project` -> pubsub topics/subscriptions / gcs buckets / bigquery tables - and what if we're reading a topic/data outside the project?). Maybe it's possible to do it a bit more cleanly 🤔

DanSimon commented on examples/catvdog/job-requirements.in:
```diff
@@ -1,6 +1,6 @@
-apache-beam[gcp]==2.22.0
+apache-beam[gcp]==2.24.0
```
since `klio-exec` also has a dependency on beam, is there a reason why we'd want to pin the version here and not defer to what `klio-exec` requires?

econchick replied:
klio-exec does not (and should not) pin the apache-beam version to anything (unless something explicitly breaks klio-exec in the future). It should be hard pinned here so folks don't deploy an upgraded version than they intended (which could easily happen (and has) if they re-build their docker image).

DanSimon commented on examples/catvdog/klio-job.yaml.example:
```diff
@@ -28,7 +29,12 @@ job_config:
         location: gs://$GCP_PROJECT-output/catvdog
+        file_suffix: .jpg
+        # as of klio-exec 0.2.0, multiple outputs aren't supported
+        # TODO: update me once it is supported!
```
is this just like a FYI or would this example job actually have multiple outputs in the future?

econchick replied:
It does have a type of multiple outputs - output could either be uploaded to `<gcs_bucket>/dog` or `<gcs_bucket>/cat`.

econchick replied:
I could re-work this a little bit so multiple outputs aren't needed 🤔

econchick replied:
actually, it isn't that easy to re-work - but I can do some custom existence checks (which would have the `skip_klio_existence_check` still set to `True` so not much different).

DanSimon reviewed (commented):
Few questions, overall looks good though

econchick commented:
@DanSimon I updated the PR - removed the python version bit, and added a custom output data existence check to handle the two subdirectories (`gs://<output-bucket>/cat` & `gs://<output-bucket>/dog`). I updated the relevant comment in the config but still left the todo in there if/when we ever support multiple outputs.

DanSimon commented on examples/catvdog/transforms.py:
```diff
@@ -125,70 +113,72 @@ def process(self, entity_id):
+            path = f"{oc.location}/{subdir}/{element}{oc.file_suffix}"
+            self._klio.logger.info(f"Checking output in {path}")
+            exists = self.gcs_client.exists(path)
+            if exists:
```
is this supposed to add something to `outputs_exist`?

econchick replied:
UGH good catch - accidentally deleted that line

DanSimon reviewed (approved):
🚀
